From f5abba68bfbc56f1b8f4a9135cfbc18115ed9004 Mon Sep 17 00:00:00 2001
From: Sebastien Guiriec <s-guiriec@ti.com>
Date: Wed, 15 Dec 2010 16:52:14 +0100
Subject: ASoC: ABE HAL: Update abe_read_gain in order to get the correct value

Update abe_read_gain() to get correct value when gain is muted or
disabled.

Change-Id: I51eb8a222716e7e6b2ba9559e2ead594e797ea1b
Signed-off-by: Sebastien Guiriec <s-guiriec@ti.com>
---
 sound/soc/omap/abe/abe_api.c |   34 +++++++++++++++++-----------------
 1 files changed, 17 insertions(+), 17 deletions(-)

Index: Kernel/sound/soc/omap/abe/abe_api.c
===================================================================
--- Kernel.orig/sound/soc/omap/abe/abe_api.c	2011-01-19 00:14:51.033322974 +0530
+++ Kernel/sound/soc/omap/abe/abe_api.c	2011-01-19 00:19:34.753321069 +0530
@@ -1702,13 +1702,6 @@
  * @p: list of port corresponding to the above gains
  *
  */
-/**
- * abe_read_gain
- * @id: name of the mixer
- * @param: list of input gains of the mixer
- * @p: list of port corresponding to the above gains
- *
- */
 abehal_status abe_read_gain(u32 id, u32 *f_g, u32 p)
 {
 	u32 mixer_target, mixer_offset, i;
@@ -1720,17 +1713,24 @@
 	mixer_target += p;
 	/* translate coef address in Bytes */
 	mixer_target <<= 2;
-	/* load the S_G_Target SMEM table */
-	abe_block_copy(COPY_FROM_ABE_TO_HOST, ABE_SMEM, mixer_target,
-		       (u32 *) f_g, sizeof(*f_g));
-	for (i = 0; i < sizeof_db2lin_table; i++) {
-		if (abe_db2lin_table[i] == *f_g)
-			goto found;
+
+	if (!abe_muted_gains_indicator[mixer_offset + p]) {
+		/* load the S_G_Target SMEM table */
+		abe_block_copy(COPY_FROM_ABE_TO_HOST, ABE_SMEM, mixer_target,
+			       (u32 *) f_g, sizeof(*f_g));
+		for (i = 0; i < sizeof_db2lin_table; i++) {
+				if (abe_db2lin_table[i] == *f_g)
+				goto found;
+		}
+		*f_g = 0;
+		return -1;
+	      found:
+		*f_g = (i * 100) + min_mdb;
+	} else {
+		/* update muted gain with new value */
+		*f_g = abe_muted_gains_decibel[mixer_offset + p];
 	}
-	*f_g = 0;
-	return -1;
-      found:
-	*f_g = (i * 100) + min_mdb;
+
 	return 0;
 }
 EXPORT_SYMBOL(abe_read_gain);
